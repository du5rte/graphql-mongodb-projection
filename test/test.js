import { graphql } from 'graphql'
import schema from './schema'

import mongoose from 'mongoose'

import UserModel from './userModel'

import mockUsers from './users'

import {
  connectMongooseAndPopulate,
  clearDB,
  disconnectMongoose
} from './utils'

let connection
let db
beforeAll(async () => {
  connection = await connectMongooseAndPopulate()
})

afterAll(async () => {
  await clearDB()
  return disconnectMongoose()
})

beforeEach(async () => {
  await clearDB()
  connection = await connectMongooseAndPopulate()
  return connection
})

describe('Mongo Setup', () => {
  it('Should populate the Users collection for testing.', async done => {
    const users = await UserModel.find({})
    expect(users).toHaveLength(mockUsers.length)
    done()
  })
})

describe('Projection Tests', function() {
  it('All Fields', async function() {
    let query = `
      {
        user(_id: "583f1607bf98f7f846e7d2d1") {
          _id
          email
          firstname
          lastname
        }
      }
    `
    return graphql(schema, query).then(result => {
      expect(result).toMatchSnapshot()
    })
  })

  it('Only selected fields', async function() {
    let query = `
      {
        user(_id: "583f1607bf98f7f846e7d2d1") {
          email
        }
      }
    `
    return graphql(schema, query).then(result => {
      expect(result).toMatchSnapshot()
    })
  })

  it('All Fields with Union Type', async function() {
    let query = `
      {
        people {
          ... on UnnamedUser {
            _id
            email
          }

          ... on User {
            _id
            email
            firstname
          }
        }
      }
    `
    return graphql(schema, query).then(result => {
      expect(result).toMatchSnapshot()
    })
  })

  it('Nested Fields with Union Type', async function() {
    let query = `
    {
      user(_id: "583f1607bf98f7f846e7d2d2") {
        _id
        email
        firstname
        lastname
        friends {
          ... on UnnamedUser {
            _id
            email
          }

          ... on User {
            _id
            firstname
          }
        }
      }
    }
    `

    return graphql(schema, query).then(result => {
      expect(result).toMatchSnapshot()
    })
  })

  it('Conditional fields', async function() {
    let query = `
    {
      user(_id: "583f1607bf98f7f846e7d2d2") {
        avatar
      }
    }
    `

    return graphql(schema, query).then(result => {
      expect(result).toMatchSnapshot()
    })
  })

  it('Nested fields', async function() {
    let query = `
    {
      user(_id: "583f1607bf98f7f846e7d2d5") {
        _id
        nested {
          level
          deep {
            deepLevel
            otherField
          }
        }
      }
    }`

    return graphql(schema, query).then(result => {
      expect(result).toMatchSnapshot()
    })
  })
})

export { connection, db }
