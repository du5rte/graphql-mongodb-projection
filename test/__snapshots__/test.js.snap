// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projection Tests All Fields 1`] = `
Object {
  "data": Object {
    "user": Object {
      "_id": "583f1607bf98f7f846e7d2d1",
      "email": "leonardodicaprio@mail.com",
      "firstname": "Leonardo",
      "lastname": "Di Caprio",
    },
  },
}
`;

exports[`Projection Tests All Fields with Union Type 1`] = `
Object {
  "data": Object {
    "people": Array [
      Object {
        "_id": "583f1607bf98f7f846e7d2d1",
        "email": "leonardodicaprio@mail.com",
        "firstname": "Leonardo",
      },
      Object {
        "_id": "583f1607bf98f7f846e7d2d2",
        "email": "bradpitt@mail.com",
        "firstname": "Brad",
      },
      Object {
        "_id": "583f1607bf98f7f846e7d2d3",
        "email": "christianbale@mail.com",
        "firstname": "Christian",
      },
      Object {
        "_id": "583f1607bf98f7f846e7d2d4",
        "email": "johndoe@mail.com",
      },
      Object {
        "_id": "583f1607bf98f7f846e7d2d5",
        "email": "aprilSmith@mail.com",
      },
    ],
  },
}
`;

exports[`Projection Tests Conditional fields 1`] = `
Object {
  "data": Object {
    "user": Object {
      "avatar": null,
    },
  },
}
`;

exports[`Projection Tests Nested Fields with Union Type 1`] = `
Object {
  "data": Object {
    "user": Object {
      "_id": "583f1607bf98f7f846e7d2d2",
      "email": "bradpitt@mail.com",
      "firstname": "Brad",
      "friends": Array [
        Object {
          "_id": "583f1607bf98f7f846e7d2d1",
          "firstname": "Leonardo",
        },
        Object {
          "_id": "583f1607bf98f7f846e7d2d3",
          "firstname": "Christian",
        },
        Object {
          "_id": "583f1607bf98f7f846e7d2d4",
          "email": "johndoe@mail.com",
        },
      ],
      "lastname": "Pitt",
    },
  },
}
`;

exports[`Projection Tests Nested fields 1`] = `
Object {
  "data": Object {
    "user": Object {
      "_id": "583f1607bf98f7f846e7d2d5",
      "nested": Object {
        "deep": Object {
          "deepLevel": "3",
          "otherField": "two",
        },
        "level": "2",
      },
    },
  },
}
`;

exports[`Projection Tests Only selected fields 1`] = `
Object {
  "data": Object {
    "user": Object {
      "email": "leonardodicaprio@mail.com",
    },
  },
}
`;
